File: javascript/leetcode/spiral-matrix-ii.js
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function(n) {
    var startN = 0;
    var endN = n - 1;
    var startM = 0;
    var endM = n - 1;
    var matrix = initializeMatrix(n);
    var position = 0;
    while (startN <= endN && startM <= endM) {
    	// go across to the right
    	for (var i = startM; i <= endM; i++) {
    		matrix[startN][i] = ++position;
    	}
    	startN++;

    	// go down
    	for (var i = startN; i <= endN; i++) {
    		matrix[i][endM] = ++position;
    	}
    	endM--;

    	// go across to the left
        if (startN <= endN) {
            for (var i = endM; i >= startM; i--) {
                matrix[endN][i] = ++position;
            }
            endN--;
        }

    	// go up
        if (startM <= endM) {
            for (var i = endN; i >= startN; i--) {
                matrix[i][startM] = ++position;
            }
            startM++;
        }
    }
    return matrix;
};

/**
 * @param {number} n
 * @return {number[][]}
 */
var initializeMatrix = function (n) {
	var matrix = [];
	for (var i = 0; i < n; i++) {
		matrix.push(new Array(n));
	}
	return matrix;
};

File: ruby/leetcode/spiral-matrix-ii.rb
# @param {Integer} n
# @return {Integer[][]}
def generate_matrix(n)
    start_n = 0
    end_n = n - 1
    start_m = 0
    end_m = n - 1
    matrix = initializeMatrix(n)
    position = 1
    while start_n <= end_n and start_m <= end_m
        # Go across the top of the matrix
        for i in start_m..(end_m) do
            matrix[start_n][i] = position
            position += 1
        end
        start_n += 1

        # Go down the right of the matrix
        for i in start_n..(end_n) do
            matrix[i][end_m] = position
            position += 1
        end
        end_m -= 1

        # Go back along the bottom
        if start_n <= end_n
            end_m.downto(start_m) do |i|
                matrix[end_n][i] = position
                position += 1
            end
            end_n -= 1
        end

        # Go up along the left
        if start_m <= end_m
            end_n.downto(start_n) do |i|
                matrix[i][start_m] = position
                position += 1
            end
            start_m += 1
        end
    end
    return matrix
end
            
def initializeMatrix(n)
    matrix = []
    for i in 0...n do
        matrix << Array.new(n)
    end
    return matrix
end
